File: src/main/java/it/unibo/crabinv/Model/Levels/LevelImpl.java
package it.unibo.crabinv.Model.Levels;

import it.unibo.crabinv.Model.Enemies.Wave;
import it.unibo.crabinv.Model.Enemies.WaveProvider;

// Adapted from MoseBarbieri's LevelLogic

public class LevelImpl implements Level{

    private final WaveProvider waveProvider;
    private Wave currentWave;

    public LevelImpl(WaveProvider waveProvider) {
        this.waveProvider = waveProvider;
        advanceWave();
    }

    @Override
    public Wave getCurrentWave() {
        return this.currentWave;
    }

    @Override
    public void advanceWave() {
        this.currentWave = waveProvider.hasMoreWaves() ? waveProvider.getNextWave() : null;
    }

    @Override
    public boolean isLevelFinished() {
        return currentWave == null;
    }
}

File: src/main/java/it/unibo/crabinv/model/levels/LevelImpl.kt
package it.unibo.crabinv.model.levels

import it.unibo.crabinv.model.entities.enemies.wave.Wave
import it.unibo.crabinv.model.entities.enemies.wave.WaveProvider

// Adapted from MoseBarbieri's LevelLogic

/**
 * Implementation of [Level].
 * @param waveProvider the provider of the wave
 */
class LevelImpl(
    private val waveProvider: WaveProvider,
) : Level {
    override var currentWave: Wave? = null
        private set

    init {
        advanceWave()
    }

    override fun advanceWave() {
        this.currentWave = if (waveProvider.hasMoreWaves()) waveProvider.getNextWave() else null
    }

    override val isLevelFinished: Boolean
        get() = currentWave == null
}
